compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -2473,9 +2472,20 @@ where
                     }
                 }
 
-                // dyn* (both fields are usize-sized)
-                ty::Dynamic(_, _, TraitObjectRepresentation::Sized) => {
-                    TyMaybeWithLayout::Ty(tcx.types.usize)
+                ty::Dynamic(_, _, ty::DynStar) => {
+                    if i == 0 {
+                        TyMaybeWithLayout::Ty(tcx.types.usize)
+                    } else if i == 1 {
+                        // FIXME(dyn-star) same FIXME as above applies here too
+                        TyMaybeWithLayout::Ty(
+                            tcx.mk_imm_ref(
+                                tcx.lifetimes.re_static,
+                                tcx.mk_array(tcx.types.usize, 3),
```

Comment:
What is this branch doing? What does the `[usize; 3]` represent?

Reply:
Oh wait, let me try and answer my own question. The first branch is if we're asking for the first field of the `dyn*`, so that's just a `usize`d data value.

The second branch is asking for the vtable, so that's a pointer to an array of pointers to methods, right?

Does this mean the 3 needs to change based on the trait and how many methods it has?

Reply:
This is the _actual_ representation of the vtable. It's not a usize, it's treated (mostly) like a `&[usize; 3]`. See the branch of `dyn` above. 

If we treat the second field like a usize like we were in the past, then llvm emits a faulty inttoptr cast.

Reply:
I think the 3 is fine because that's the same thing we have for `&dyn` above. It's probably not completely correct since we're technically reading past the end of the vtable for user-provided methods, but that's not something unique to `dyn*`.

Reply:
See:

```
 Warning: naively computing the number of entries in the
vtable by counting the methods on the trait + methods on
all parent traits does not work, because some methods can
be not object safe and thus excluded from the vtable.
Increase this counter if you tried to implement this but
failed to do it without duplicating a lot of code from
other places in the compiler: 2
```